src/plugins/intel_gpu/src/plugin/infer_request.cpp

```diff
@@ -857,17 +857,27 @@ void InferRequest::allocate_dev_mem_if_needed(InferenceEngine::BlobMap& device_m
         // so we don't need to allocate new memory
         can_skip_allocation |= same_host_mem(impl_mem, src_ptr);
         // Or if blob has any type except usm_host - in that case explicit copy will be performed anyway
-        can_skip_allocation |= impl_mem->get_allocation_type() != cldnn::allocation_type::usm_host;
+        // Or if blob has usm_host type and lockable memory is expected by impl
+        can_skip_allocation |= !need_lockable_mem ? impl_mem->get_allocation_type() != cldnn::allocation_type::usm_host
+                                                  : impl_mem->get_allocation_type() == cldnn::allocation_type::usm_host;
+        // In case of lockable memory we need to keep updated device's usm_host memory buffer with
+        // user's blob to avoid incorrect behaviour if user will call set_blob() with
+        // the following sequence (usm_host, system_host, usm_host, system_host...)
+        can_skip_allocation &= !need_lockable_mem ? true : users_blobs_matching[blob_name] == user_blob;
```

Comment:
Is it possible that has_device_blob == true, but blob_name is not in users_blobs_matching? If yes, you'll get an exception here.

Reply:
It looks like it is not possible, but I have added condition for that